twitch and vimeo support to videostream component

VanshulB opened the pull request:
Extending the videostream component, functionality to view livestream conference when using Twitch and Vimeo RTMP URI, This pull request includes changes to fixes #64 that allow the embedding of the Twitch and Vimeo players, enabling users to view the livestream conference directly on the website.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/RC4Conferences?pullRequest=93) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/RC4Conferences?pullRequest=93) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/RC4Conferences?pullRequest=93) it.</sub>

VanshulB commented:
Here is a working demo of the feature using the Twitch RTMP URI.

https://user-images.githubusercontent.com/98731471/214879471-72e79c6f-6740-4889-9b9f-287abe40f870.mov


Note The embedded player will not load when using the Vimeo RTMP URI (Start Free Demo), as the viewing of that event is private and is limited to your viewing only, as seen in the screenshot. In order to broadcast using Vimeo and have that stream be visible to the public, you must subscribe to their premium service. As a result, using the Vimeo Free Demo won't function because the event's viewership isn't public or password-protected.

<img width="1680" alt="Screenshot 2023-01-26 at 7 56 55 PM" src="https://user-images.githubusercontent.com/98731471/214880406-f90dbd23-0c67-47d7-a014-3fbc44d9d5d0.png">

Dnouv commented on app/components/conferences/dayOfEvent/mainstage/videostreamer.js:
```diff
@@ -41,50 +41,100 @@ export default function Videostreamer(props) {
+                channel: props.src,
+                layout: "video",
+                autoplay: false,
+                parent: ["127.0.0.1", "localhost"],
```
Could you let me know what this does? Since if we are going to use this in production, it will definitely have a different IP.
If it is not necessary, please remove this, or if it is, try to find a way to handle the production and development server.
Thank you!

VanshulB replied:
From what I've read and understood from the Twitch docs, The `parent` key in the player object specifies the parent origin for embedding Twitch videos, and its value is an array of strings that lists the permitted IP addresses/Domains of the embedded content. It is necessary to add the `parent` key, otherwise the Twitch player won't show the livestream and throw some error related to not the right kind of environment to show livestream. In the production environment, we should replace the `parent` key's value with the actual array of domain, where the embedded Twitch livestream is expected to be displayed. Thus It's necessary to correctly set the parent value in-order to ensure that the embedded content is only displayed in the intended and trusted environment and to prevent any security vulnerabilities.

Dnouv replied:
Ok, thank you. Now I understand so it is like whitelisting the domain to show the Twitch stream.

Hmm, this is tricky, as you have seen we design things in RC4Conferences as a component, so we could not ask any user to edit a file before editing edit. Could you make this read from an environment variable? 
FYI, we have an env variable available during the build which is `NEXT_PUBLIC_API_URL=http://localhost:3000` (this is the URL where the NextJS client is running), so if you could use it in the parent key, that would be great. 

Thank you!

Dnouv commented on docs/conferences/dayofevent/README.md:
````diff
@@ -47,6 +47,11 @@ In order to provide a better stream to the attendees, based on their proximate d
 ```
 NEXT_PUBLIC_IPINFO_TOKEN="token from ipinfo"
 ```
+
+If you are using __Twitch__ RTMP URI then in order to view the stream you need to add specify your Twitch username in the `src` prop of the Videostream tag in the Mainstage component `/app/components/conferences/dayOfEvent/mainstage/Mainstage.js`.
+
````
Nice, you wrote the documentation. 

One addition is required could you please also add more detailed documentation of this `VideoStreamer` component in `https://github.com/RocketChat/RC4Conferences/tree/main/docs/components`? Since you have been working on this feature, so you are definitely much more familiar with it.

Use [this](https://github.com/RocketChat/RC4Conferences/wiki/Event-Management-Components#event-scheduler) component as a reference; once you have done this, I will copy and those to our wiki page. Oh, and for the "Preview" part, please feel free to add a video.

Thank you!

VanshulB replied:
Hello @Dnouv 
I have added the wiki documentation of the videostream component, Since this is my first time writing documentation, kindly review it and point out any mistakes I've made. I've included the wiki documentation for the videostream component to the repository at `/docs/components/videostreamer/README.md`.

Dnouv replied:
Thank you! I will read it, from what I have read till now, it is written pretty well, nice!

Dnouv commented:
Please fix the styling of the embedded Twitch stream. It is not responsive and gets overlapped when the chat window is opened.

On Mobile Screen:
<img width="300" src="https://user-images.githubusercontent.com/61188295/215268359-7b20d55e-22cb-4132-b043-98cebeece24b.png" alt="3000-vanshulb-rc4conferences-g7tkdlvicxe ws-us84 gitpod io_conferences_mainstage_9ddffcbb(iPhone 12 Pro)"/>

With Chat Window Open:

![3000-vanshulb-rc4conferences-g7tkdlvicxe ws-us84 gitpod io_conferences_mainstage_9ddffcbb (2)](https://user-images.githubusercontent.com/61188295/215268420-7bd8ae7e-486a-474e-86ca-654cc27c0923.png)

Dnouv commented:
@VanshulB, thanks for the PR; the component works perfectly fine. It just needs some styling and documentation changes, so once these are rectified we will be good to merge this. Please have a look at the comments.
Thank you!

VanshulB commented:
Hello @Dnouv 
I made the embedded player responsive and corrected the design; for your convenience, I've added a demo video. Please let me know if I made any mistakes.


https://user-images.githubusercontent.com/98731471/215465495-6af31280-f9e8-4c64-9dbc-06eaa546b793.mp4

Dnouv commented on docs/conferences/dayofevent/README.md:
```diff
@@ -47,6 +48,14 @@ In order to provide a better stream to the attendees, based on their proximate d
+
+If you are using __Twitch__ RTMP URI then in order to view the stream you need to add the `service` prop as `service={"twitch"}` your Twitch username in the `src` prop of the Videostream tag in the Mainstage component `/app/components/conferences/dayOfEvent/mainstage/Mainstage.js`.
+
+Else if you are using __Vimeo__ RTMP URI then then in order to view the stream you need to add the `service` prop as `service={"vimeo"}` your Vimeo live event link in the `src` prop of the Videostream tag in the Mainstage component `/app/components/conferences/dayOfEvent/mainstage/Mainstage.js`. It will be in the format of `https://vimeo.com/webinars/events/{Random-String}/embed`
```
Great work on adding the responsiveness, I hope you have done the same for Vimeo also. Please fix the `parentKey`, and we will be good to merge.
Thank you!

VanshulB replied:
Yes I have added responsiveness to the Vimeo player as well, in the parent key I have added the check if the NODE_ENV is production the parent key will be the domain on which RC4Conferences is deployed otherwise it will be localhost.

Dnouv commented on app/components/conferences/dayOfEvent/mainstage/videostreamer.js:
```diff
@@ -10,6 +10,9 @@ export default function Videostreamer(props) {
     "Thanks for joining in! Please start the stream."
   );
 
+  const parentDevelopement = ["127.0.0.1", "localhost"];
+  const parentProduction = ["https://conf.rceng.shop"];
```
Hey @VanshulB 
Thanks for the fix, but that's not how the scalability is handled. Think about the following use cases:
1. I am doing the development on a VM, which definitely will have `172.xx.yy.z` something like this IP instead of the `127.0.0.1`
2. For production, we are not the only users going to use RC4Conferences; suppose you have a community website (`community.vanshul.home`) and integrated this component into it, then this won't work.

Please think carefully about this cases. Thank you!

VanshulB replied:
Hey @Dnouv 

In order to handle scalability, we have used environment variables to set the domain name/ ip dynamically instead of hardcoding them into the codebase! have added the documentation for adding the environment variable if we are developing in a virtual machine or while using in the production environment we can specify the domain on which our website is gonna host. I hope that's what you wanted! Please tell any suggestions.
Thank you!

Dnouv commented:
Hey @VanshulB 

Thanks for the PR! And congrats 🎉 !!!
